### PR TITLE
Refactor forced cleanup of state tracker and move it to UI thread executor

### PR DESCRIPTION
**Pull request by axi0mX**
Refactoring. No functional changes.

1) Move forced cleanup of state tracker to UI thread executor.

I removed the other commits and they'll be added to another PR.

### CODE REVIEW

File: EarlGrey/Synchronization/GREYUIThreadExecutor.h [hunk 1]
```diff
@@ -130,4 +130,13 @@ typedef void(^GREYExecBlock)(void);
                          block:(GREYExecBlock)execBlock
                          error:(__strong NSError **)error;
 
+/**
+ *  Drains the UI thread and waits for both the UI and idling resources to idle, for up to
+ *  @c seconds, after which GREYAppStateTracker is forcefully cleared if it is not idle.
+ *
+ * @param seconds Amount of time to wait for the UI and idling resources to idle before forcefully
```
Comment (tirodkar) [inline EarlGrey/Synchronization/GREYUIThreadExecutor.h]: Align.

**Comment by tirodkar** [comment]
Could we get some descriptions as to what the issue is that the CL is addressing and how it solves it? Maybe a way to test it as well? Thanks a lot!!

File: EarlGrey/Synchronization/GREYUIThreadExecutor.h (hunk 1 above)
Comment (axi0mX) [inline EarlGrey/Synchronization/GREYUIThreadExecutor.h]: Aligned

**Comment by axi0mX** [comment]
This pull request only refactors some code, no functional changes.

File: EarlGrey/Synchronization/GREYUIThreadExecutor.h [hunk 2]
```diff
@@ -130,4 +130,13 @@ typedef void(^GREYExecBlock)(void);
...
                          error:(__strong NSError **)error;
 
+/**
+ *  Drains the UI thread and waits for both the UI and idling resources to idle, for up to
+ *  @c seconds, after which GREYAppStateTracker is forcefully cleared if it is not idle.
+ *
+ *  @param seconds Amount of time to wait for the UI and idling resources to idle before forcefully
+ *                 clearing GREYAppStateTracker.
+ */
+- (void)performForcedCleanUpAfterTimeout:(CFTimeInterval)seconds;
```
Comment (tirodkar) [inline EarlGrey/Synchronization/GREYUIThreadExecutor.h]: Since we're exposing this method, it would be preferable to write some UnitTests for it :)

Comment (axi0mX) [inline EarlGrey/Synchronization/GREYUIThreadExecutor.h]: I made it private. Let me know if you think we should expose it.

**Comment by tirodkar** [comment]
@axi0mX, this needs to be rebased after the dispatch_tracker change that was added by @sid-github

**Comment by axi0mX** [comment]
Done

File: EarlGrey/Synchronization/GREYUIThreadExecutor.m [hunk 3]
```diff
@@ -276,6 +257,34 @@ - (BOOL)grey_isTrackingIdlingResource:(id<GREYIdlingResource>)idlingResource {
 #pragma mark - Private
 
 /**
+ *  Register the specified @c resource to be checked for idling before executing test actions.
+ *  A strong reference is held to @c resource until it is deregistered using
+ *  @c grey_deregisterIdlingResource. It is safe to call this from any thread.
+ *
+ *  @param resource The idling resource to register.
+ */
+- (void)grey_registerIdlingResource:(id<GREYIdlingResource>)resource {
```
Comment (khandpur) [inline EarlGrey/Synchronization/GREYUIThreadExecutor.m]: we plan to clean this up and expose it as public API. there are quite a few usages of this internally so please don't rename this.

Comment (axi0mX) [inline EarlGrey/Synchronization/GREYUIThreadExecutor.m]: Removed.

**Pull request merged**